### PR TITLE
reva/frontend: Add capabilities to indicate attributes that are read-only

### PR DESCRIPTION
**Pull request by Excds**
This adds capabilities indicating if certain attributes are read-only.

Companion PR: https://github.com/cs3org/reva/pull/3789

### CODE REVIEW

**Comment by mmattel** [comment]
Please add some text to readme.md to reflect the changes on the envvar  side.
I can help finetuning after a first writeup.

File: services/frontend/README.md
Comment (mmattel) [inline services/frontend/README.md:46]:
```suggestion
## Libregraph Service Interactions

A lot of user management is done via a standardized libregraph API. Depending on how the system is configured, there might be some attributes for users that an ocis instance admin user can't change because of properties
coming from an external LDAP server, or similar. This can be the case when the ocis admin is not the LDAP admin. To make life easier for admin users, there are hints as capabilites telling the frontend which attributes are read-only or not, so they can be shown in the frontend differently. To configure these hints the following environment variables are available:

-   FRONTEND_READONLY_ATTRIBUTES_ACCOUNT_ENABLED: Defaults to false
-   FRONTEND_READONLY_ATTRIBUTES_DISPLAY_NAME: Defaults to true
-   FRONTEND_READONLY_ATTRIBUTES_GIVEN_NAME: Defaults to true
-   FRONTEND_READONLY_ATTRIBUTES_ID: Defaults to true
-   FRONTEND_READONLY_ATTRIBUTES_MAIL: Defaults to true
-   FRONTEND_READONLY_ATTRIBUTES_ON_PREMISES_SAM_ACCOUNT_NAME: Defaults to true
-   FRONTEND_READONLY_ATTRIBUTES_SURNAME: Defaults to true
-   FRONTEND_READONLY_ATTRIBUTES_QUOTA: Defaults to false
```

File: services/frontend/pkg/config/config.go
Comment (mmattel) [inline services/frontend/pkg/config/config.go:173]:
```suggestion
	AccountEnabled           bool `yaml:"account_enabled" env:"FRONTEND_READONLY_ATTRIBUTES_ACCOUNT_ENABLED" desc:"Flag to indicate if account_enabled attribute is read-only. Defaults to false."`
	DisplayName              bool `yaml:"display_name" env:"FRONTEND_READONLY_ATTRIBUTES_DISPLAY_NAME" desc:"Flag to indicate if display_name attribute is read-only. Defaults to true."`
	GivenName                bool `yaml:"given_name" env:"FRONTEND_READONLY_ATTRIBUTES_GIVEN_NAME" desc:"Flag to indicate if given_name attribute is read-only. Defaults to true."`
	ID                       bool `yaml:"id" env:"FRONTEND_READONLY_ATTRIBUTES_ID" desc:"Flag to indicate if id attribute is read-only. Defaults to true."`
	Mail                     bool `yaml:"mail" env:"FRONTEND_READONLY_ATTRIBUTES_MAIL" desc:"Flag to indicate if mail attribute is read-only. Defaults to true."`
	OnPremisesSamAccountName bool `yaml:"on_premises_sam_account_name" env:"FRONTEND_READONLY_ATTRIBUTES_ON_PREMISES_SAM_ACCOUNT_NAME" desc:"Flag to indicate if on_premises_sam_account_name attribute is read-only. Defaults to true."`
	Surname                  bool `yaml:"surname" env:"FRONTEND_READONLY_ATTRIBUTES_SURNAME" desc:"Flag to indicate if surname attribute is read-only.  Defaults to true."`
	Quota                    bool `yaml:"quota" env:"FRONTEND_READONLY_ATTRIBUTES_QUOTA" desc:"Flag to indicate if quota attribute read-only. Defaults to false."`
```

**Review by mmattel — APPROVED** [review]
LGTM from a docs pov, but needs a dev reviewer approval too

**Comment by micbar** [comment]
@mmattel this has changed in the reva implementation. We will have only one config value and a list of strings.

**Comment by Excds** [comment]
> @mmattel this has changed in the reva implementation. We will have only one config value and a list of strings.

@micbar In the ocis implementation, each attribute is still an environment variable.

File: services/frontend/pkg/config/defaults/defaultconfig.go
Comment (kulmann) [inline services/frontend/pkg/config/defaults/defaultconfig.go:92]: Just for my understanding, no clever way of finding out if the built in libre idm is used and in that case having all these attributes writable automatically? 😢

**Comment by micbar** [comment]
> > @mmattel this has changed in the reva implementation. We will have only one config value and a list of strings.
> 
> @micbar In the ocis implementation, each attribute is still an environment variable.

That surprises me, it would mean we need a code change for every new attribute.

**Review by rhafer — CHANGES_REQUESTED** [review]

File: services/frontend/pkg/config/defaults/defaultconfig.go
Comment (kulmann) [reply services/frontend/pkg/config/defaults/defaultconfig.go]: Unresolved the conversation, not convinced, yet. sorry.

oCIS single binary comes with a builtin IdP where the attributes are in fact writable. From an admin user perspective it's a huge step backwards if they now need to set environment variables to get the shipped artifact (ocis single binary) into it's "default state" of having writable fields.

For external IdPs it's of course a matter of correct configuration whether or not the fields are writable. But for the single binary that we ship you would cripple functionality.

File: services/frontend/pkg/config/defaults/defaultconfig.go
Comment (kulmann) [reply services/frontend/pkg/config/defaults/defaultconfig.go]: I'm having the same concerns

File: services/frontend/README.md
Comment (mmattel) [inline services/frontend/README.md:33]:
```suggestion
which takes a comma separated list of attributes. The default value for graph service attributes is `display_name,given_name,id,mail,on_premises_sam_account_name,surname`.
```
Make the string (list of items) better readable by using ticks.

File: services/frontend/pkg/config/defaults/defaultconfig.go
Comment (Excds) [reply services/frontend/pkg/config/defaults/defaultconfig.go]: Updated.

File: services/frontend/pkg/config/defaults/defaultconfig.go
Comment (Excds) [reply services/frontend/pkg/config/defaults/defaultconfig.go]: @rhafer What's your thoughts on this?

Comment (kobergj) [reply services/frontend/pkg/config/defaults/defaultconfig.go]: I also want to join the party :smile: 

Wouldn't `ReadOnlyAttributes` anyways be empty in the single binary case? That would mean all a writable or not?

File: services/frontend/pkg/config/config.go
Comment (kobergj) [inline services/frontend/pkg/config/config.go:47]: Can we make this a `[]string`? That's what it is.

**Review by kobergj — CHANGES_REQUESTED** [review]

File: services/frontend/pkg/config/defaults/defaultconfig.go
Comment (kulmann) [reply services/frontend/pkg/config/defaults/defaultconfig.go]: Yes, that's basically what @rhafer wrote in his PR review. I agree, empty string as default for single binary would be correct and would solve my concerns.

If an external IdP is used it's most likely that the env variable must be filled and if forgotten the UI suggests wrong behaviour. But that's still better than setting most fields to read only and having `standard single binary` users confused. Happy to go with that solution.

Comment (rhafer) [reply services/frontend/pkg/config/defaults/defaultconfig.go]: > @rhafer What's your thoughts on this?

That's why we should have and empty default here. So I guess this discussion is resolved, right?

> Wouldn't ReadOnlyAttributes anyways be empty in the single binary case? That would mean all are writable or not?

Yes. The setting should be empty by default. See also the other conversation, below.

**Comment by micbar** [comment]
@Excds Can we make the adjustments and merge this today to unblock the frontend team?

**Comment by micbar** [comment]
AFAICT there is no reva update inside this PR, this will be needed to make the changes work.

File: services/frontend/pkg/config/config.go
Comment (Excds) [reply services/frontend/pkg/config/config.go]: Done.

File: services/frontend/pkg/revaconfig/config.go
Comment (kobergj) [inline services/frontend/pkg/revaconfig/config.go:69]: Is this necessary? Can't we just assign `cfg.ReadOnlyUserAttributes` directly? It shouldn't care if it is `nil`

**Review by kobergj — CHANGES_REQUESTED** [review]
One more question

File: services/frontend/README.md
Comment (mmattel) [inline services/frontend/README.md:33]:
```suggestion
A lot of user management is made via the standardized libregraph API. Depending on how the system is configured, there might be some user attributes that an ocis instance admin can't change because of properties coming from an external LDAP server, or similar. This can be the case when the ocis admin is not the LDAP admin. To ease life for admins, there are hints as capabilites telling the frontend which attributes are read-only to enable a different optical representation like being grayed out. To configure these hints, use the environment variable `FRONTEND_READONLY_USER_ATTRIBUTES`, which takes a comma separated list of attributes.
```
Just a small improvement after re-reading.

File: services/frontend/README.md
Comment (mmattel) [inline services/frontend/README.md:29]: I am not happy with the section title because it is hard to understand what is meant.
What about:

`Define Read-Only Attributes`

File: services/frontend/pkg/revaconfig/config.go
Comment (Excds) [reply services/frontend/pkg/revaconfig/config.go:69]: I added that because I thought it would be nicer for the frontend people if they'd just have a list to iterate over and not have to check for nil. I'm totally fine with both.

File: services/frontend/README.md
Comment (mmattel) [reply services/frontend/README.md:29]: done 😄

File: services/frontend/pkg/revaconfig/config.go
Comment (kobergj) [reply services/frontend/pkg/revaconfig/config.go:69]: Fine with me :+1:

**Review by kobergj — APPROVED** [review]

**Review by rhafer — APPROVED** [review]

**Pull request merged**
